message/validation/alan/consensus_validation.go

```diff
@@ -0,0 +1,427 @@
+package msgvalidation
+
+// consensus_validation.go contains methods for validating consensus messages
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"time"
+
+	"github.com/attestantio/go-eth2-client/spec/phase0"
+	specqbft "github.com/bloxapp/ssv-spec/alan/qbft"
+	spectypes "github.com/bloxapp/ssv-spec/alan/types"
+	"github.com/bloxapp/ssv-spec/types"
+
+	"github.com/bloxapp/ssv/protocol/v2/qbft/roundtimer"
+	ssvtypes "github.com/bloxapp/ssv/protocol/v2/types"
+)
+
+func (mv *messageValidator) validateConsensusMessage(
+	signedSSVMessage *spectypes.SignedSSVMessage,
+	committee []spectypes.OperatorID,
+	validatorIndices []phase0.ValidatorIndex,
+	receivedAt time.Time,
+) (*specqbft.Message, error) {
+	ssvMessage := signedSSVMessage.GetSSVMessage()
+
+	if len(ssvMessage.Data) > maxConsensusMsgSize {
+		e := ErrSSVDataTooBig
+		e.got = len(ssvMessage.Data)
+		e.want = maxConsensusMsgSize
+		return nil, e
+	}
+
+	consensusMessage, err := specqbft.DecodeMessage(ssvMessage.Data)
+	if err != nil {
+		e := ErrUndecodableMessageData
+		e.innerErr = err
+		return nil, e
+	}
+
+	mv.reportConsensusMessageMetrics(signedSSVMessage, consensusMessage, committee)
+
+	if err := mv.validateConsensusMessageSemantics(signedSSVMessage, consensusMessage, committee); err != nil {
+		return consensusMessage, err
+	}
+
+	state := mv.consensusState(signedSSVMessage.GetSSVMessage().GetID())
+
+	if err := mv.validateQBFTLogic(signedSSVMessage, consensusMessage, committee, receivedAt, state); err != nil {
+		return consensusMessage, err
+	}
+
+	if err := mv.validateQBFTMessageByDutyLogic(signedSSVMessage, consensusMessage, validatorIndices, receivedAt, state); err != nil {
+		return consensusMessage, err
+	}
+
+	for i := range signedSSVMessage.GetSignature() {
+		operatorID := signedSSVMessage.GetOperatorIDs()[i]
+		signature := signedSSVMessage.GetSignature()[i]
+
+		if err := mv.verifySignature(ssvMessage, operatorID, signature); err != nil {
+			return consensusMessage, err
+		}
+	}
+
+	mv.updateConsensusState(signedSSVMessage, consensusMessage, state)
+
+	return consensusMessage, nil
+}
+
+func (mv *messageValidator) validateConsensusMessageSemantics(
+	signedSSVMessage *spectypes.SignedSSVMessage,
+	consensusMessage *specqbft.Message,
+	committee []spectypes.OperatorID,
+) error {
+	signers := signedSSVMessage.GetOperatorIDs()
+	quorumSize, _ := ssvtypes.ComputeQuorumAndPartialQuorum(len(committee))
+	if len(signers) > 1 {
+		if consensusMessage.MsgType != specqbft.CommitMsgType {
+			e := ErrNonDecidedWithMultipleSigners
+			e.got = len(signers)
+			return e
+		}
+
+		if uint64(len(signers)) < quorumSize {
+			e := ErrDecidedNotEnoughSigners
+			e.want = quorumSize
+			e.got = len(signers)
+			return e
+		}
```

Comment:
I also limit it to be <= committee size. Is it possible to be added here?

Reply:
Yes, it's `ErrTooManySigners` in `belongsToCommittee`